homeassistant/components/weatherflow_cloud/coordinator.py

```diff
@@ -1,39 +1,169 @@
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        rest_api: WeatherFlowRestAPI,
+        stations: StationsResponseREST,
+    ) -> None:
         """Initialize global WeatherFlow forecast data updater."""
-        self.weather_api = WeatherFlowRestAPI(api_token=api_token)
+
         super().__init__(
             hass,
-            LOGGER,
-            name=DOMAIN,
+            rest_api,
+            stations,
             update_interval=timedelta(seconds=60),
+            always_update=True,
         )
 
     async def _async_update_data(self) -> dict[int, WeatherFlowDataREST]:
-        """Fetch data from WeatherFlow Forecast."""
+        """Update rest data."""
         try:
-            async with self.weather_api:
-                return await self.weather_api.get_all_data()
+            async with self._rest_api:
+                return await self._rest_api.get_all_data()
         except ClientResponseError as err:
             if err.status == 401:
                 raise ConfigEntryAuthFailed(err) from err
             raise UpdateFailed(f"Update failed: {err}") from err
+
+    def get_station(self, station_id: int):
+        """Return station for id."""
+        return self.data[station_id]
+
+    def get_station_name(self, station_id: int):
+        """Return station name for id."""
+        return self.data[station_id].station.name
+
+
+class WeatherFlowCloudDataCallbackCoordinator[
+    T: EventDataRapidWind | WebsocketObservation,
+    M: RapidWindListenStartMessage | ListenStartMessage,
+    C: RapidWindWS | ObservationTempestWS,
+](BaseWeatherFlowCoordinator[dict[int, T | None]]):
+    """A Generic coordinator to handle Websocket connections.
+
+    This class takes 3 generics - T, M, and C.
+    T - The type of data that will be stored in the coordinator.
+    M - The type of message that will be sent to the websocket API.
+    C - The type of message that will be received from the websocket API.
+    """
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        rest_api: WeatherFlowRestAPI,
+        websocket_api: WeatherFlowWebsocketAPI,
+        stations: StationsResponseREST,
+        listen_request_type: type[M],
+        event_type: EventType,
+    ) -> None:
+        """Initialize Coordinator."""
+
+        super().__init__(hass=hass, rest_api=rest_api, stations=stations)
+
+        self._event_type = event_type
+        self.websocket_api = websocket_api
+        self._listen_request_type = listen_request_type
+
+        # configure the websocket data structure
+        self._ws_data: dict[int, dict[int, T | None]] = {
+            station: {device: None for device in devices}
+            for station, devices in self.stations.station_device_map.items()
+        }
+
+    async def _generic_callback(self, data: C):
+        device_id = data.device_id
+        station_id = self.device_to_station_map[device_id]
+        self._ws_data[station_id][device_id] = getattr(data, "ob", data)
```

Comment:
why the getattr?

Reply:
![image](https://github.com/user-attachments/assets/f5b6dc03-0084-45d4-a225-cd4781a89642)

(Couldn't resist)

Reply:
The actual answer - we are receiving two types of data across the Websocket:

`RapidWindWS` which looks like:

```json

{
    "device_id": 211522,
    "serial_number": "ST-00084623",
    "type": "rapid_wind",
    "hub_sn": "HB-00068684",
    "ob": [
      1709141614,
      2.47,
      198
    ]
  }
```

And an `ObservationTempestWS` event - which we parse differently

Reply:
But then this is not very generic if we just "try" to get data or fallback to another method

Reply:
Does this work for you?

```python
        if isinstance(data, RapidWindWS):
            processed_data = data.ob
        elif isinstance(data, ObservationTempestWS):
            processed_data = data
        else:
            LOGGER.warning("Unknown message type received: %s", type(data))
            return
```

I remember reading/hearing once that `is instance` was a VERY computationally slow call - so I tried to avoid it... but I guess this shoudl be fine.